### ISSUE TITLE
Provide textual way to search and run commands in Mastodon

### ISSUE THREAD

**Issue by stefanhahmann** [mastodon-sc/mastodon#309]
# User Story

In Fiji the, there is a search box that can be used to search and run FIJI commands.

![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/b3bb043c-93e1-4731-ae65-2d4908a99bdc)

This provides a convenient way to find and access functions for users, which does not require to search within the menus. It would be cool to have a similar mechanism also within Mastodon, since there are also a lot of functions in the core and also in the plugins, which are difficult to explore.

I could imagine three options how to provide this function:

1. Add a search box to Mastodon windows in the lower right corner just as in Fiji to search Mastodon and run Mastodon commands: 
![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/c0d41a4c-b420-4a3b-8481-1c338143056c)
2. Add a search box to the main Mastodon panel: 
![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/bda0cbb7-14c4-4842-88a0-73957f0f937b)
3. Add a new "Test" button in the Keymap settings that would run the currently filtered/selected command 
![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/64216c7f-b1dc-4917-b5f3-f0e7716b6466)

**Comment by stefanhahmann** [comment]
I think this branch goes very much into the direction to what was asked by the user who approached me with this idea.

This would be my feedback after testing:
* It would be nice, if there was an alternative way to access the command finder. With only the `CTRL + Shift + F` shortcut, it would be a bit hidden. I think, if there was some additional visual search bar, users would benefit even more from the existence of the Mastodon command finder
* The list of available commands seems incomplete:
  * E.g., if opened from BDV, I cannot set the 2d drag rotate in the command finder, while in the dialog `File > Preferences > KeyMap`, I can: 
![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/b8cd2679-5f04-4faf-bed7-d3d701c38e86)
  * It is unclear to me how commands from plugins could be added to the command finder
* The command finder shows duplicates, because for some commands multiple short cuts are defined. For such situations, it would be nice, if only one row would be used that shows all shorts cuts. Cf. e.g. this example from IntelliJ: 
![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/921da4ce-8092-4c48-9e71-8034de1f8d0c)
* The Workflow e.g. `CTRL + Shift + F` > `Fix Project Path` > `Run` > `Cancel` should maybe leave the command finder open, since users may want to run some other commands
* A `Double Click` on an action should start the action as well

**Comment by stefanhahmann** [comment]
I will make the version of this branch available for the user who had the idea for this feature and ask him for feedback as well.

**Comment by stefanhahmann** [comment]
> Yes, this is true. I am still a bit unsure though. I think that if we add too many widgets we will loose data content and confuse the new users. I need to think and use it.

I see your point. I would still vote for it, since it helps also new users.

> It's actually on purpose. What you mention here are `Behaviors`, they require something with the mouse (drag, click etc). The command finder only work for `Actions` that can just run as is.

I see. How can actions from plugins be added?

**Comment by tinevez** [comment]
> I see. How can actions from plugins be added?

I realize that I collect actions only from one `Actions` instance. In Mastodon, they are spread over several instances, sorted by category. I need to rewrite this to include several instances, including the `Plugins` one.

**Comment by stefanhahmann** [comment]
> Mrmlml not such a fan of letting it open. Maybe we could make this configurable?

Yes. Maybe this is an option. A checkbox in the command finder dialog "keep open". It could be off by default.

> I realize that I collect actions only from one Actions instance. In Mastodon, they are spread over several instances, sorted by category. I need to rewrite this to include several instances, including the Plugins one.

Sounds good.

**Comment by tinevez** [comment]
Can you test again?
- the plugin actions should be supported
- installed also on the main window
- common commands are merged in the table
- double-click works

**Comment by tinevez** [comment]
- commands are updated when the user edits them
- the 'manual' descriptions can be added

**Comment by stefanhahmann** [comment]
> Can you test again?

Yes.

> the plugin actions should be supported

I still have difficulties with this. E.g. I was looking in the command finder for some actions from the Trees Management, but can't see them: 

![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/752ccd33-8aef-4fb2-ba27-eb5d333ea563)

In the Preferences dialog, I can see them however:

![grafik](https://github.com/mastodon-sc/mastodon/assets/10515534/c33359c9-2b7f-4fc1-a5a3-959cd8a6326b)

Is there something that needs to be adapted on the side of the plugins?
 
> installed also on the main window

Works nicely!
 
> common commands are merged in the table

Works nicely!

>double-click works

Works nicely.

**Comment by tinevez** [comment]
I think you need to run the command finder on the window that can accept the commands to see them. 
For instance the tree management commands you mention are only visible in `trackscheme` and `bdv` context, so they will appear only in these windows. 
If you want them to appear everywhere you need to add them to the `mastodon` context

**Comment by stefanhahmann** [comment]
Yes, I can confirm that it works exactly like this. Thanks for the explanation

**Comment by jgirstmair** [comment]
Hi JY and Stefan! I tested the latest version of the command finder, which I find extremely useful (I am the user who asked for it 😅 inspired by Fiji's quick search bar). I couldn't find any issues and was already happy to discover some new commands like 'select all links' that I haven't been aware of. I will keep testing it and come back here in case I notice something unexpected or if I have further suggestions. But right now I think it could stay like it is. One minor thing I agree with Stefan: It would be nice to have the option somewhere that the command finder stays open after running a command. I had a feeling Fiji's quick search bar used to behave like this, but it got changed. Right now it also closes after executing any command. Is there a specific reason why this was changed and why it should be avoided in Mastodon's command finder? Cheers
